proposals/nnnn-mutex.md

```diff
@@ -0,0 +1,480 @@
+
+## Motivation
+
+In concurrent programs, protecting shared mutable state is one of the core fundamental problems to ensuring reading and writing data is done in an explainable fashion. Synchronizing access to shared mutable state is not a new problem in Swift. We've introduced many features to help protect mutable data. Actors are a good default go-to solution for protecting mutable state because it isolates the stored data in its own domain. At any given point in time, only one task will be executing "on" the actor, and have exclusive access to it. Multiple tasks cannot access state protected by the actor at the same time, although they may interleave execution at potential suspension points (indicated by `await`). In general, the actor approach also leans itself well to code organization, since the actor's state, and operations on this state are logically declared in the same place: inside the actor.
+
+Not all code may be (or want to) able to adopt actors. Reasons for this can be very varied, for example code may have to execute synchronously without any potential for other tasks interleaving with it. Or the `async` effect introduced on methods may prevent legacy code which cannot use Swift Concurrency from interacting with the protected state.
+
+Whatever the reason may be, it may not be feasible to use an actor. In such cases, Swift currently is missing standardized tools to offer developers ensure proper synchronization in their concurrent data-structures. Many Swift programs opt to use ad-hoc implementations of a mutal exclusion lock, or a mutex. A mutex is a simple to use synchronization primitive to help protect shared mutable data by ensuring that a single execution context has exclusive access to the related data. The main issue is that there isn't a single standardized implementation for this synchronization primitive resulting in everyone needing to roll their own.
+
+## Proposed solution
+
+We propose a new type in the Standard Library Synchronization module: `Mutex`. This type will be a wrapper over a platform-specific mutex primitive, along with a user-defined mutable state to protect. Below is an example use of `Mutex` protecting some internal data in a class usable simultaneously by many threads:
+
+```swift
+class FancyManagerOfSorts {
+  let cache = Mutex<[String: Resource]>([:])
+  
+  func save(_ resource: Resouce, as key: String) {
+    cache.withLock {
+      $0[key] = resource
+    }
+  }
+}
+```
+
+Use cases for such a synchronized type are common. Another common need is a global cache, such as a dictionary:
+
+```swift
+let globalCache = Mutex<[MyKey: MyValue]>([:])
+```
+
+## Detailed design
+
+### Underlying System Mutex Implementation
+
+The `Mutex` type proposed is a wrapper around a platform's implementation.
+
+* macOS, iOS, watchOS, tvOS, visionOS:
```

Comment:
What's the behaviour when compiling for WASM? Will Mutex be available on Embedded?

Reply:
WASM has support for atomic operations and even a `wait` and `wake` so we can reasonably build something like Linux's implementation on WASM. I'm unsure if those features are only available for X version of WASM however. Embedded doesn't include the Synchronization module at all right now, so Mutex will not be available yet/ever.

Reply:
It looks like this support is implemented in some places, but the actual proposal hasn't been accepted yet...?

Reply:
Ah looks like it's in phase 4 of the proposal which is standardize the thing https://github.com/WebAssembly/proposals

Reply:
@MaxDesiatov points out that this feature is implemented in all browsers at this point though https://webassembly.org/features/

Reply:
cc @kateinoigakukun

Reply:
LLVM also implements the [`*.atomic.*` instructions](https://github.com/WebAssembly/threads/blob/main/proposals/threads/Overview.md) in their backend and it's enabled only when a specific CPU feature (`-matomics`) is enabled. So for both Embedded/non-Embedded Wasm targets, it would be best to have `Mutex` implementation when the feature is enabled. (Also if we can provide no-op implementation when the feature is not enabled, it can reduce a large amount of `#if` in the use-site)

We can have separate Swift SDKs for each feature set configuration, so we don't need to care Wasm's versioning here.

Anyway, I can help enable this module for WebAssembly targets.